serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt

```diff
@@ -78,44 +78,120 @@ class MatrixRooms(private val msession: MatrixSession, val message: String) : Ma
         return MatrixLogin("Closing session, returning to the login prompt for now\n", MatrixClient())
     }
 }
-data class SharedUiMessage(
-    val sender: String,
-    val message: String,
-    val id: String,
-    val timestamp: Long,
-    val url: String? = null
+open class SharedUiMessage(
+    open val sender: String,
+    open val message: String,
+    open val id: String,
+    open val timestamp: Long,
+    open val replied_event: String = "",
+    open val edited_event: String = ""
+
 )
+class SharedUiImgMessage(
+    override val sender: String,
+    override val message: String,
+    override val id: String,
+    override val timestamp: Long,
+    val url: String
+) : SharedUiMessage(sender,message,id,timestamp)
+class SharedUiAudioMessage(
+    override val sender: String,
+    override val message: String,
+    override val id: String,
+    override val timestamp: Long,
+    val url: String
+) : SharedUiMessage(sender,message,id,timestamp)
+
 class MatrixChatRoom(private val msession: MatrixSession, val room_id: String, val name: String) : MatrixState() {
+    val edits: MutableMap<String,SharedUiMessage> = mutableMapOf()
```

Comment:
I think the shared UI should do the edit combination for the UI, not exposing the mutable edit map, just the same list of messages it always has, just pre-edited. This would keep the UI thinner, and it can use extra fields on the SharedUiMessage to display it specially or allow jumping to the referenced event. Gives edit handling for free for simple GUIs too!

Reply:
Agreed